### PR TITLE
Add MediaPipe Hands implementation

### PR DESCRIPTION
**Pull request by ahmedsabie**
This implementation of hand detector wraps the existing [npm mediapipe implementation](https://www.npmjs.com/package/@mediapipe/hands)

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/839)

### CODE REVIEW

**Review by lina128 — APPROVED** [review]
This is great, LGTM overall. Just some nits. Please wait for Ivan, Valentin and Michael's review.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-:-MkIYm8N8yCuY12DbsI0:by40jfs)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @ahmedsabie)

___
*[hand-detection/README.md, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIRl5IE06I--SA1hYd:-MkIRl5IE06I--SA1hYe:b-7eskw3) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L35)):*
> ```
> poses
> ```
hands
___
*[hand-detection/README.md, line 36 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIRpSECeIgVdS668yD:-MkIRpSECeIgVdS668yE:bqx1w3e) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L36)):*
> ```
> For single-hand models
> ```
Is there a single-hand model? Sugg: remove this sentence.
___
*[hand-detection/README.md, line 38 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkISRO1CiNK2cwEAogY:-MkISRO1CiNK2cwEAogZ:b-y5t2hk) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L38)):*
> ```
> an array of keypoints
> ```
Is there also a keypoints3D?
___
*[hand-detection/README.md, line 40 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIS8E9-k5Tmr89Fl64:-MkIS8E9-k5Tmr89Fl65:bmedamm) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L40)):*
> ```
> x, y, 
> ```
Is there a z dimension?
___
*[hand-detection/README.md, line 61 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkISZhV5R_yVk2kdAqF:-MkISZhV5R_yVk2kdAqG:b8gjehx) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L61)):*
> ```
> Refer to each model's documentation for specific configurations for the model
> and their performance.
> ```
We should have our own MediaPipe API readme under the src/mediapipe folder, because their web solution API is different than our API. Remove this for now.
___
*[hand-detection/src/mediapipe/constants.ts, line 28 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkITQBJ4Y-DZwsEqGuv:-MkITQBK1Fy8xFnaMo0C:b-c2q2ro) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/constants.ts#L28)):*
> ```
> true
> ```
default to false.
___
*[hand-detection/src/mediapipe/detector.ts, line 39 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIV0i8-JrJoKBIb_dZ:-MkIV0i8-JrJoKBIb_d_:bh1rp7l) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L39)):*
> ```
> private selfieMode = false;
> ```
Hi Michael, I forgot what this field is about, seems we always set it to false, do we need it? Is there a default if we don't pass in this field to handsSolution? @mhays-google 
___
*[hand-detection/src/mediapipe/detector.ts, line 54 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIVPTSEJAeZYhPjARi:-MkIVPTSEJAeZYhPjARj:b-s188ug) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L54)):*
> ```
> || 2
> ```
You shouldn't need it, because it's already validated.
___
*[hand-detection/src/mediapipe/detector.ts, line 66 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIWolI3bQBjE-MPvQM:-MkIWolI3bQBjE-MPvQN:b-z0obpf) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L66)):*
> ```
> keypoints:
> ```
Is there also a keypoints3D?
___
*[hand-detection/src/mediapipe/detector.ts, line 99 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIX8YICXPvDx5kMUdE:-MkIX8YICXPvDx5kMUdF:b-y6vytq) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L99)):*
> ```
> Defaults to true
> ```
Can we default to false? I think the more common use case is video.
___
*[hand-detection/src/mediapipe/detector.ts, line 101 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIXN-p1-mJmM5p2MDv:-MkIXN-p1-mJmM5p2MDw:buh2qlz) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L101)):*
> ```
>    *       implementation. Image input types are assumed to be static images,
>    *       and video inputs are assumed to be non static images.
> ```
I thought the flag can override the underlying assumption.
___
*[hand-detection/src/mediapipe/detector.ts, line 125 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIXgIbBO359NMmNBLC:-MkIXgIc4IOuNYKR4s4C:b6yrige) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L125)):*
> ```
> this.handsSolution.reset();
> ```
when you reset, should you also reset all the member variables that store states? Such as height, width, and hands.
___
*[hand-detection/src/mediapipe/detector.ts, line 136 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkITxtsD2OP9GpWHwUT:-MkITxtsD2OP9GpWHwUU:b-q68e8t) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L136)):*
> ```
> ModelConfig object
> ```
An object
___
*[hand-detection/src/mediapipe/detector.ts, line 144 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIU6gp02AJA-9t18Pu:-MkIU6gp02AJA-9t18Pv:bb5jiin) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L144)):*
> ```
> result
> ```
nit: detector
___
*[hand-detection/src/mediapipe/README.md, line 34 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIYOmv1jilIXQwwuOH:-MkIYOmv1jilIXQwwuOI:bpej758) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/README.md#L34)):*
> ```
> Hands
> ```
hand-detection
___
*[hand-detection/test_data/hands.jpg, line 0 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIYfO069vC2wbftuM6:-MkIYfO069vC2wbftuM7:bhqr95q) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/test_data/hands.jpg#L0)):*
Make sure test_data/ is in .npmignore

**Review by ahmedsabie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/839)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @ahmedsabie, @bazarevsky, @ivan-grishchenko, @lina128, and @mhays-google)

___
*[hand-detection/README.md, line 35 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIRl5IE06I--SA1hYd:-MkIwFjA9G_qQAzpK_6P:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L35)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> poses
> ```
hands
</blockquote></details>

Done.
___
*[hand-detection/README.md, line 36 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIRpSECeIgVdS668yD:-MkIwM38CwRqFMQCc4j3:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L36)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> For single-hand models
> ```
Is there a single-hand model? Sugg: remove this sentence.
</blockquote></details>

Done.
___
*[hand-detection/README.md, line 38 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkISRO1CiNK2cwEAogY:-MkIwMYn2j51EjNjr9ed:b-bajka) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L38)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> an array of keypoints
> ```
Is there also a keypoints3D?
</blockquote></details>

not yet
___
*[hand-detection/README.md, line 40 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIS8E9-k5Tmr89Fl64:-MkIwQdG03VFbXO2FJsK:b-bajka) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L40)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> x, y, 
> ```
Is there a z dimension?
</blockquote></details>

not yet
___
*[hand-detection/README.md, line 61 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkISZhV5R_yVk2kdAqF:-MkIwaUO77cNilyjjKuP:b-pu31hg) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L61)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> Refer to each model's documentation for specific configurations for the model
> and their performance.
> ```
We should have our own MediaPipe API readme under the src/mediapipe folder, because their web solution API is different than our API. Remove this for now.
</blockquote></details>

the readme for src/mediapipe is included in this PR, should I keep this sentence and hyperlink below?
___
*[hand-detection/src/mediapipe/constants.ts, line 28 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkITQBJ4Y-DZwsEqGuv:-MkIxNeDCVIsHGHnMmZr:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/constants.ts#L28)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> true
> ```
default to false.
</blockquote></details>

Done.
___
*[hand-detection/src/mediapipe/detector.ts, line 54 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIVPTSEJAeZYhPjARi:-MkIxjRv5FgZUyoPtVUQ:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L54)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> || 2
> ```
You shouldn't need it, because it's already validated.
</blockquote></details>

Done.
___
*[hand-detection/src/mediapipe/detector.ts, line 66 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIWolI3bQBjE-MPvQM:-MkIxjzDBvkPrB7kVfjh:b-bajka) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L66)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> keypoints:
> ```
Is there also a keypoints3D?
</blockquote></details>

not yet
___
*[hand-detection/src/mediapipe/detector.ts, line 99 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIX8YICXPvDx5kMUdE:-MkIxrto36f2rwPCXUhi:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L99)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> Defaults to true
> ```
Can we default to false? I think the more common use case is video.
</blockquote></details>

Done.
___
*[hand-detection/src/mediapipe/detector.ts, line 101 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIXN-p1-mJmM5p2MDv:-MkIy1cG1pcFQpRV0tcm:b-xjjt69) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L101)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
>    *       implementation. Image input types are assumed to be static images,
>    *       and video inputs are assumed to be non static images.
> ```
I thought the flag can override the underlying assumption.
</blockquote></details>

Mediapipe API doesn't support this flag, it uses the input type image or video to decide, I can remove it from the mediapipe config and add it specifically to the TFJS implementation later
___
*[hand-detection/src/mediapipe/detector.ts, line 125 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIXgIbBO359NMmNBLC:-MkIyyuX04_SfzD5XKbS:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L125)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> this.handsSolution.reset();
> ```
when you reset, should you also reset all the member variables that store states? Such as height, width, and hands.
</blockquote></details>

Done.
___
*[hand-detection/src/mediapipe/detector.ts, line 136 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkITxtsD2OP9GpWHwUT:-MkIz1CXBiOLPwzs7ec9:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L136)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> ModelConfig object
> ```
An object
</blockquote></details>

Done.
___
*[hand-detection/src/mediapipe/README.md, line 34 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIYOmv1jilIXQwwuOH:-MkIzFzZ0nPaU8lXBJQ0:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/README.md#L34)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> Hands
> ```
hand-detection
</blockquote></details>

Done.
___
*[hand-detection/test_data/hands.jpg, line  at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIYfO069vC2wbftuM6:-MkIzH974_GO8Fp1k6TI:b-7j5sih) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/test_data/hands.jpg#L)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Make sure test_data/ is in .npmignore
</blockquote></details>

great catch added an npmignore

**Review by lina128 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/839)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @ahmedsabie, @bazarevsky, @ivan-grishchenko, and @lina128)

___
*[hand-detection/README.md, line 61 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkISZhV5R_yVk2kdAqF:-MkNuRrcA930cCGH7MkM:beiqdp6) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/README.md#L61)):*
<details><summary><i>Previously, ahmedsabie wrote…</i></summary><blockquote>

the readme for src/mediapipe is included in this PR, should I keep this sentence and hyperlink below?
</blockquote></details>

Yes, hyperlink to src/mediapipe/README.md
___
*[hand-detection/src/mediapipe/detector.ts, line 101 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIXN-p1-mJmM5p2MDv:-MkNudoQ1h07arX5wkvz:b2nhnp1) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/mediapipe/detector.ts#L101)):*
<details><summary><i>Previously, ahmedsabie wrote…</i></summary><blockquote>

Mediapipe API doesn't support this flag, it uses the input type image or video to decide, I can remove it from the mediapipe config and add it specifically to the TFJS implementation later
</blockquote></details>

SG.

File: hand-detection/src/mediapipe/detector.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        }
+      }
+    });
+  }
+
+  private translateOutput(landmarks: hands.NormalizedLandmarkList): Keypoint[] {
+    return landmarks.map((landmark, i) => ({
+                           x: landmark.x * this.width,
+                           y: landmark.y * this.height,
+                           z: landmark.z,
```
Comment (bazarevsky) [inline hand-detection/src/mediapipe/detector.ts:79]: I see some inconsistency here. We usually scale Z landmarks similarly to x to preserve the same ratio in 3D. Here we clearly loose this scale information for z. Please check.

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/839)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @ahmedsabie, @bazarevsky, @ivan-grishchenko, and @lina128)

___
*[hand-detection/src/mediapipe/detector.ts, line 79 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkEM2koQrtkGLiWnQcG-r2-79:-MkOW4WVD2DbcjF-uw7l:bcy9fdu) ([raw file](https://github.com/tensorflow/tfjs-models/blob/b76d2284631a48411126533523121b59a9c6b797/hand-detection/src/mediapipe/detector.ts#L79)):*
<details><summary><i>Previously, bazarevsky (v.bazarevsky) wrote…</i></summary><blockquote>

I see some inconsistency here. We usually scale Z landmarks similarly to x to preserve the same ratio in 3D. Here we clearly loose this scale information for z. Please check.
</blockquote></details>

I'm actually curious, does this z have any meaning? Note, this is not the 3D model.

**Review by ahmedsabie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/839)* status: :shipit: complete! 2 of 1 approvals obtained (waiting on @ahmedsabie, @bazarevsky, @ivan-grishchenko, @lina128, and @mhays-google)

___
*[hand-detection/src/create_detector.ts, line 36 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkIeIZX3QRwvQEdxGDX:-MkPTGPu07Dbpl7DLnGc:b-482j9y) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb54c0f25dc2068dee11eb250f1071212952af7f/hand-detection/src/create_detector.ts#L36)):*
<details><summary><i>Previously, mhays-google (Michael Hays) wrote…</i></summary><blockquote>

Making sure you want the type coercion here...
</blockquote></details>

It's currently needed for compilation since the tfjs runtime is not added yet but it's better to have an error documenting that we will support it
___
*[hand-detection/src/mediapipe/detector.ts, line 79 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/839#-MkEM2koQrtkGLiWnQcG-r2-79:-MkPTolLAPuOd-Ahs0y0:b-l9kp5i) ([raw file](https://github.com/tensorflow/tfjs-models/blob/b76d2284631a48411126533523121b59a9c6b797/hand-detection/src/mediapipe/detector.ts#L79)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

I'm actually curious, does this z have any meaning? Note, this is not the 3D model.
</blockquote></details>

Fixed the scaling, also sorry I thought there wasn't a z since the test case didn't check its value, but I went back to the mediapipe docs and it says "z represents the landmark depth with the depth at the wrist being the origin, and the smaller the value the closer the landmark is to the camera." I updated the readme with this as well

**Pull request merged**